packages/core/src/app/pg-core/db/chain-db.ts

```diff
@@ -1,24 +1,523 @@
+          : computeParent(sibling, node)
+        await this.addMerkleTreeNode(blockNumber, parentIndex, parent)
+      }
+
+      // Stop, don't need to delete anything else.
+      return
+    }
+
+    // We don't have the node, but we have children that can be used to
+    // re-generate it. Need to recursively delete any of these children.
+
+    // We're at a leaf node so we can't go any deeper.
+    const treeHeight = await this.getBlockTreeHeight(blockNumber)
+    if (nodeIndex > 2 ** treeHeight) {
+      return
+    }
+
+    // Compute left and right children.
+    const leftChildIndex = getLeftChildIndex(nodeIndex)
+    const rightChildIndex = getRightChildIndex(nodeIndex)
+
+    // Delete the children.
+    await this.removeMerkleTreeNode(blockNumber, leftChildIndex)
+    await this.removeMerkleTreeNode(blockNumber, rightChildIndex)
+  }
+
+  /**
+   * Smart method for removing Merkle proof nodes stored for a specific
+   * transaction. Checks the ensure that deletion of the nodes wouldn't impact
+   * other transaction proofs.
+   * @param blockNumber Block number to delete nodes from.
+   * @param leafIndex Index of the leaf to delete proof nodes for.
+   */
+  public async removeMerkleProofNodes(
+    blockNumber: number,
+    leafIndex: number
+  ): Promise<void> {
+    // Get the leaf indices of all other transactions in the same block.
+    const otherLeafIndices: number[] = (await this.getAllLeafIndices(
+      blockNumber
+    )).filter((otherLeafIndex) => {
+      return otherLeafIndex !== leafIndex
+    })
+
+    // Need the tree height to compute sibling/parent nodes.
+    const treeHeight = await this.getBlockTreeHeight(blockNumber)
+
+    // We don't want to delete any nodes that other transactions still need.
+    // Compute the siblings for *all* other stored transactions so we're not
+    // deleting anything critical.
+    let allOtherSiblingIndices: number[] = []
+    for (const otherLeafIndex of otherLeafIndices) {
+      const otherSiblingIndices = getMerkleSiblingIndices(
+        otherLeafIndex,
+        treeHeight
+      )
+      allOtherSiblingIndices = allOtherSiblingIndices.concat(
+        otherSiblingIndices
+      )
+    }
+
+    // Now figure out which sibling indices we can safely get rid of by finding
+    // anything that other nodes don't need. This is a safe operation because
+    // `removeMerkleTreeNode` will insert any parent that could've been
+    // computed by the removed node.
+    const siblingIndices = getMerkleSiblingIndices(leafIndex, treeHeight)
+    for (const siblingIndex of siblingIndices) {
+      if (!allOtherSiblingIndices.includes(siblingIndex)) {
+        await this.removeMerkleTreeNode(blockNumber, siblingIndex)
+      }
+    }
+  }
+
+  /**
+   * Creates an inclusion proof for a given transaction.
+   * @param transaction Transaction to create a proof for.
+   * @returns the inclusion proof for that transaction.
+   */
+  public async getInclusionProof(
```

Comment:
Not sure how this actually plays with the Merkle sum tree. Is this method independent of the underlying Merkle tree?